tools/ansible/tasks/wsu/main.yaml

```diff
@@ -20,25 +20,160 @@
+    - name: Get kubernetes changelog url
+      shell: |
+        echo "https://github.com/kubernetes/kubernetes/blob/release-"\
+        "{{ kubernetes_version_without_patch.stdout }}/CHANGELOG/CHANGELOG-"\
+        "{{ kubernetes_version_without_patch.stdout }}.md"
+      register: changelog
+
+    - name: Get kube node binary SHA
+      shell: |
+        curl -s "{{ changelog.stdout }}" | \
+        grep -A 1 "{{ kubernetes_version.stdout }}" | \
+        grep -A 1 node-windows | tail -n 1 | sed -e 's/.*<code>\(.*\)<\/code>.*/\1/'
+      register: kube_node_sha
+
+    - name: Set kubelet location
+      set_fact:
+        # Example of kubelet download url: "https://dl.k8s.io/v1.16.2/kubernetes-node-windows-amd64.tar.gz"
+        kubelet_location: "https://dl.k8s.io/{{ kubernetes_version.stdout }}/kubernetes-node-windows-amd64.tar.gz"
+
+    - name: Get kubernetes minor version
+      shell: "echo {{ kubernetes_version_without_patch.stdout }} | cut -d '.' -f2"
+      register: kubernetes_minor_version
+
+    # This translates a kubernetes minor version to an OpenShift version. This works under the assumption that each
+    # OpenShift release corresponds with a kubernetes release. This is being done this way, and not with oc get
+    # clusterversion, as OpenShift CI doesn't have the actual version attached to its clusters, instead replacing it
+    # with 0.0.1 and information about the release creation date
+    # TODO Track the validity of this script when OpenShift version is bumped and the corresponding Kubernetes version is not
+    - name: Get cluster version
+      shell: |
+        MINORVERSION={{ kubernetes_minor_version.stdout }}
+        BASEK8SVERSION=16
+        BASEOPENSHIFTVERSION=3
+        if [ \"$MINORVERSION\" -lt \"$BASEK8SVERSION\" ]; then
+        echo unsupported kubernetes version; exit 1
+        fi
+        VERSIONINCREMENTS=$((MINORVERSION - BASEK8SVERSION))
+        echo 4.$((BASEOPENSHIFTVERSION + VERSIONINCREMENTS))
+        exit 0
+      register: cluster_version
+
+    # curl -s https://api.github.com/repos/openshift/windows-machine-config-bootstrapper/releases gets the list of all
+    # releases, we use the cluster version from previous ansible task to select the release we need.
+    # TODO Ensure latest release is available for every new OpenShift version
+    - name: Get release
+      shell: |
+        RELEASES="$(curl -s https://api.github.com/repos/openshift/windows-machine-config-bootstrapper/releases)"
+        TAGNAME="$(echo $RELEASES | jq '.[]' | jq '.tag_name' | \
+        grep -i {{ cluster_version.stdout }} |\
+        sort -r | awk 'NR==1')"
+        echo $RELEASES | \
+        jq '.[] | select (.tag_name == '$TAGNAME')'
+      register: release
+
+    - name: Check release object
+      fail:
+        msg: "Cannot find the latest release"
+      when: release is not defined or release.stdout == ""
```

Comment:
What is the Ansible standard for using this vs `failed_when`?

Reply:
Doesn't make much difference except `fail` gives us a custom error message. I must have missed this as a part of my revisions - I am removing this module and adding a `failed_when` in the task above with the same condition